### PR TITLE
[Guide]: App shell UX with service workers and streams

### LINKED ISSUE #2868: content: App Shell UX with Service Workers and Streams [2020-06-12]
* [Draft](https://docs.google.com/document/d/1IR0YXEOuZ466iCiegSsGbsSr-vaM-nQX0Ke993_DQu4/edit)
* Authors: @demianrenzulli & @jeffposnick 
* CC: @petele 
* Reviewer: @jpmedley

### PR DESCRIPTION
**Pull request by demianrenzulli**
Fixes #2868

Fourth guide of the "Advanced Patterns" section of Web.dev/reliable.

An new entry for this post should be created at src/site/_data/paths/reliable.js, after [the PR that starts the series](https://github.com/GoogleChrome/web.dev/pull/3248/files) is merged.

cc // @kaycebasques @jeffposnick

### CODE REVIEW

File: src/site/content/en/reliable/app-shell-ux-sw/index.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+---
```
Comment (kaycebasques) [inline src/site/content/en/reliable/app-shell-ux-sw/index.md:1]: Please change the URL to `app-shell-ux-with-service-workers`

File: src/site/content/en/reliable/app-shell-ux-sw/index.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+[Single-page app (SPA)](https://en.wikipedia.org/wiki/Single-page_application) is an architectural pattern in which the browser runs JavaScript code to update the existing page when the user visits a different section of the site, as opposed to loading an entire new page.
+
+This means that the web app doesn't perform an actual page reload. The [History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API) is used instead to navigate back and forth through the user's history and manipulate the contents of the history stack.
+
+Using this type of architecture can provide an [app shell UX](https://developers.google.com/web/fundamentals/architecture/app-shell) that's fast, reliable, and usually consumes less data when navigating.
+
+In multi-page apps (MPAs) each time a user navigates to a new URL, the browser progressively renders HTML specific to that page. This means a full page refresh every time you visit a new page.
+
+While both are equally valid models to use, you might want to bring some of the benefits of the app-shell UX of SPAs to your existing MPA site.
```
Comment (kaycebasques) [inline src/site/content/en/reliable/app-shell-ux-sw/index.md:20]:
```suggestion
While both are equally valid models to use, you might want to bring some of the benefits of the app shell UX of SPAs to your existing MPA site.
```

File: src/site/content/en/reliable/app-shell-ux-sw/index.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+In this article we'll analyze how you can achieve an SPA-like architecture in multi-page apps by combining partials, service workers, and streams.
+
+## Production case
+
+[DEV](https://dev.to/) is a platform where software developers write articles, take part in discussions, and build their professional profiles.
+
+Their architecture is a multi-page app based on traditional backend templating through Ruby on Rails. Their team was interested in some of the benefits of an app shell model, but didn't want to undertake a major architectural change or move away from their original tech stack.
```
Comment (kaycebasques) [inline src/site/content/en/reliable/app-shell-ux-sw/index.md:27]:
```suggestion
[DEV](https://dev.to/) is a platform where software developers write articles, take part in discussions, and build their professional profiles.

<figure class="w-figure">
  <img class="w-screenshot w-screenshot--filled" 
       src="homepage.jpg" 
       alt="A screenshot of https://dev.to">
</figure>

Their architecture is a multi-page app based on traditional backend templating through Ruby on Rails. Their team was interested in some of the benefits of an app shell model, but didn't want to undertake a major architectural change or move away from their original tech stack. 
```

Please add this picture to your branch as `homepage.jpg`

homepage

File: src/site/content/en/reliable/app-shell-ux-sw/index.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
+---
+layout: post
+title: App Shell UX with service workers and streams
```
Comment (kaycebasques) [inline src/site/content/en/reliable/app-shell-ux-sw/index.md:3]:
```suggestion
title: App shell UX with service workers and streams
```
Our site convention is to use sentence case

File: src/site/content/en/reliable/app-shell-ux-sw/index.md [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+Using this type of architecture can provide an [app shell UX](https://developers.google.com/web/fundamentals/architecture/app-shell) that's fast, reliable, and usually consumes less data when navigating.
+
+In multi-page apps (MPAs) each time a user navigates to a new URL, the browser progressively renders HTML specific to that page. This means a full page refresh every time you visit a new page.
+
+While both are equally valid models to use, you might want to bring some of the benefits of the app-shell UX of SPAs to your existing MPA site.
+In this article we'll analyze how you can achieve an SPA-like architecture in multi-page apps by combining partials, service workers, and streams.
+
+## Production case
+
+[DEV](https://dev.to/) is a platform where software developers write articles, take part in discussions, and build their professional profiles.
```
Comment (kaycebasques) [inline src/site/content/en/reliable/app-shell-ux-sw/index.md:25]:
```suggestion
[DEV](https://dev.to/) is a community where software developers write articles, take part in discussions, and build their professional profiles.
```
I think `community` is more accurate description of what they're all about?

File: src/site/content/en/reliable/app-shell-ux-sw/index.md [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+</figure>
+
+The key element of this solution is the usage of [streams](https://developers.google.com/web/updates/2016/06/sw-readablestreams), which enables [incremental creations and updates](https://streams.spec.whatwg.org/#intro) of data sources. The Streams API also provides an interface for reading or writing asynchronous chunks of data, only a subset of which might be available in memory at any given time.
+This way, the header of the page can be rendered as soon as it's picked from the cache, while the rest of the content is being fetched from the network. As a result, the navigation experience is so fast that users don't perceive an actual page refresh, only the new content (the body) being updated.
+
+The resulting UX is similar to the app shell UX pattern of SPAs, implemented on a MPA site.
+
+{% Aside %} 
+The previous section contains a quick summary of dev.to's solution. For a more detailed explanation check out their [blog post](https://dev.to/devteam/instant-webpages-and-terabytes-of-data-savings-through-the-magic-of-service-workers-1mkc) on this topic.
```
Comment (kaycebasques) [inline src/site/content/en/reliable/app-shell-ux-sw/index.md:45]:
```suggestion
The previous section contains a quick summary of DEV's solution. For a more detailed explanation check out their [blog post](https://dev.to/devteam/instant-webpages-and-terabytes-of-data-savings-through-the-magic-of-service-workers-1mkc) on this topic.
```
This is how they appear to brand their company name

File: src/site/content/en/reliable/app-shell-ux-sw/index.md (hunk 3 above)
Comment (demianrenzulli) [reply src/site/content/en/reliable/app-shell-ux-sw/index.md:27]: This is very nice, thank you for taking the time to take the screenshot.

**Pull request merged**